### ISSUE TITLE
Enum list in variables not working after migration to v5 (breaking)

### ISSUE THREAD

**Issue by draganjovanovic1** [graphql-dotnet/graphql-dotnet#3091]
## Description

Argument of enum list throws parse exception when passed through variables. It works when passed in query directly.

```
{
  "errors": [
    {
      "message": "GraphQL.Validation.InvalidVariableError: Variable '$types[0]' is invalid. Unable to convert 'PASSENGER_TRANSFER' to 'TransportationTypeEnum'\n ---> System.InvalidOperationException: Unable to convert 'PASSENGER_TRANSFER' to the scalar type 'TransportationTypeEnum'\n   at GraphQL.Types.ScalarGraphType.ThrowValueConversionError(Object value) in /_/src/GraphQL/Types/Scalars/ScalarGraphType.cs:line 235\n   at GraphQL.Types.EnumerationGraphType.ParseValue(Object value) in /_/src/GraphQL/Types/Scalars/Enumeration/EnumerationGraphType.cs:line 81\n   at GraphQL.Validation.ValidationContext.<GetVariableValue>g__ParseValueScalar|49_1(ScalarGraphType scalarGraphType, GraphQLVariableDefinition variableDef, VariableName variableName, Object value) in /_/src/GraphQL/Validation/ValidationContext.cs:line 253\n   --- End of inner exception stack trace ---\n   at GraphQL.Validation.ValidationContext.<GetVariableValue>g__ParseValueScalar|49_1(ScalarGraphType scalarGraphType, GraphQLVariableDefinition variableDef, VariableName variableName, Object value) in /_/src/GraphQL/Validation/ValidationContext.cs:line 253\n   at GraphQL.Validation.ValidationContext.<GetVariableValue>g__ParseValue|49_0(IGraphType type, GraphQLVariableDefinition variableDef, VariableName variableName, Object value, IVariableVisitor visitor) in /_/src/GraphQL/Validation/ValidationContext.cs:line 222\n   at GraphQL.Validation.ValidationContext.<GetVariableValue>g__ParseValueList|49_2(ListGraphType listGraphType, GraphQLVariableDefinition variableDef, VariableName variableName, Object value, IVariableVisitor visitor) in /_/src/GraphQL/Validation/ValidationContext.cs:line 273\n   at GraphQL.Validation.ValidationContext.<GetVariableValue>g__ParseValue|49_0(IGraphType type, GraphQLVariableDefinition variableDef, VariableName variableName, Object value, IVariableVisitor visitor) in /_/src/GraphQL/Validation/ValidationContext.cs:line 222\n   at GraphQL.Validation.ValidationContext.GetVariableValues(IVariableVisitor visitor) in /_/src/GraphQL/Validation/ValidationContext.cs:line 156",
      "locations": [
        {
          "line": 1,
          "column": 11
        }
      ],
      "extensions": {
        "code": "INVALID_VALUE",
        "codes": [
          "INVALID_VALUE",
          "INVALID_OPERATION"
        ],
        "number": "5.8"
      }
    }
  ]
}
```

Multiline exception, for better readability: 
```
GraphQL.Validation.InvalidVariableError: Variable '$types[0]' is invalid. Unable to convert 'PASSENGER_TRANSFER' to 'TransportationTypeEnum'
 ---> System.InvalidOperationException: Unable to convert 'PASSENGER_TRANSFER' to the scalar type 'TransportationTypeEnum'
   at GraphQL.Types.ScalarGraphType.ThrowValueConversionError(Object value) in /_/src/GraphQL/Types/Scalars/ScalarGraphType.cs:line 235
   at GraphQL.Types.EnumerationGraphType.ParseValue(Object value) in /_/src/GraphQL/Types/Scalars/Enumeration/EnumerationGraphType.cs:line 81
   at GraphQL.Validation.ValidationContext.<GetVariableValue>g__ParseValueScalar|49_1(ScalarGraphType scalarGraphType, GraphQLVariableDefinition variableDef, VariableName variableName, Object value) in /_/src/GraphQL/Validation/ValidationContext.cs:line 253
   --- End of inner exception stack trace ---
   at GraphQL.Validation.ValidationContext.<GetVariableValue>g__ParseValueScalar|49_1(ScalarGraphType scalarGraphType, GraphQLVariableDefinition variableDef, VariableName variableName, Object value) in /_/src/GraphQL/Validation/ValidationContext.cs:line 253
   at GraphQL.Validation.ValidationContext.<GetVariableValue>g__ParseValue|49_0(IGraphType type, GraphQLVariableDefinition variableDef, VariableName variableName, Object value, IVariableVisitor visitor) in /_/src/GraphQL/Validation/ValidationContext.cs:line 222
   at GraphQL.Validation.ValidationContext.<GetVariableValue>g__ParseValueList|49_2(ListGraphType listGraphType, GraphQLVariableDefinition variableDef, VariableName variableName, Object value, IVariableVisitor visitor) in /_/src/GraphQL/Validation/ValidationContext.cs:line 273
   at GraphQL.Validation.ValidationContext.<GetVariableValue>g__ParseValue|49_0(IGraphType type, GraphQLVariableDefinition variableDef, VariableName variableName, Object value, IVariableVisitor visitor) in /_/src/GraphQL/Validation/ValidationContext.cs:line 222
   at GraphQL.Validation.ValidationContext.GetVariableValues(IVariableVisitor visitor) in /_/src/GraphQL/Validation/ValidationContext.cs:line 156"
```

### Steps to reproduce

Create enum type e.g. `TransportationTypeEnum`. Create query with argument of `ListGraphType<TransportationTypeEnum>` and pass value through variables.

Enum as parameter:
```
query plansInPeriod($types: [TransportationTypeEnum!]){
  transportationPlans(
    from: "2022-01-03"
    to: "2022-04-20"
    companyId: "bd29c46144934a199729625b6c716975"
    skip: 0
    take: 10 
    types: $types
  ) {
    totalCount
  }
}
```

Variables (does not work, throws exception, see above):
```
{
  "types": ["PASSENGER_TRANSFER"]
}
```

Variables (works when not in array !?!?):
```
{
  "types": "PASSENGER_TRANSFER"
}
```

Enum in query (works):
```
query plansInPeriod($types: [TransportationTypeEnum!]){
  transportationPlans(
    from: "2022-01-03"
    to: "2022-04-20"
    companyId: "bd29c46144934a199729625b6c716975"
    skip: 0
    take: 10 
    types: [PASSENGER_TRANSFER]
  ) {
    totalCount
  }
}
```

### Expected result

Argument should work from variables (it used to work in 4.7.2). 

### Actual result

Argument parsing fails.

### Environment

macOS Monterey 12.3.1, MacBook Pro M1Max

**Comment by Shane32** [comment]
And also the `types` argument definition for `plansInPeriod` - thanks!

**Comment by draganjovanovic1** [comment]
Sure... It's F# and across multiple files but I collected all together here, no modifications.

```fsharp
    [<RequireQualifiedAccess>]
    module TransportationTypeKind =
        let [<Literal>] PassengerTransfer = "PASSENGER_TRANSFER"
        let [<Literal>] CrewSafeguarding = "CREW_SAFEGUARDING"

    let addEnumValue (x: EnumerationGraphType) value desc =
        x.Add (name = value, value = value, description = desc)

    type TransportationTypeEnumGraphType () as x =
        inherit EnumerationGraphType ()

        do
            x.Name <- typeName x
            x.Description <- "Transportation type"

            addEnumValue x TransportationTypeKind.PassengerTransfer "Passenger transfer"
            addEnumValue x TransportationTypeKind.CrewSafeguarding "Crew safeguarding"
```

Basically I put `PASSENGER_TRANSFER` string in both `name` and `value` properties.

**Comment by Shane32** [comment]
I updated the test in #3092 to have a similar call to `Add`... still I can't find any issues.

I don't know F# well enough -- is `TransportationTypeKind.PassengerTransfer` a constant or some type like an enum?  If you put the literal string "PASSENGER_TRANSFER" on that "addEnumValue" line does it work the same?

Can you define the `value` argument of `addEnumValue` to ensure it is a string -- maybe like this:
```fsharp
    let addEnumValue (x: EnumerationGraphType) (value: string) desc =
        x.Add (name = value, value = value, description = desc)
```

Again, I don't know F# ...

**Comment by draganjovanovic1** [comment]
It is string. Those are string constants. 

```fsharp
let [<Literal>] PassengerTransfer = "PASSENGER_TRANSFER"
```

Because `PASSENGER_TRANSFER` is string literal it is inferred as string (no type definition required but it is the same even if you explicitly define it).

**Comment by draganjovanovic1** [comment]
Unfortunately, didn't help. It works just fine in `v4.7.1` but not in `v5.1.1` which is blocker for me to upgrade.

Basically the code is the same as:

```fsharp
    type TransportationTypeEnumGraphType () as x =
        inherit EnumerationGraphType ()

        do
            x.Name <- typeName x
            x.Description <- "Transportation type"

            x.Add (name = "PASSENGER_TRANSFER", value = "PASSENGER_TRANSFER", description = "Passenger transfer")
            x.Add (name = "CREW_SAFEGUARDING", value = "CREW_SAFEGUARDING", description = "Crew safeguarding")
```

**Comment by Shane32** [comment]
Can you post the definition of `typeName`?

**Comment by Shane32** [comment]
Still have not been able to reproduce the problem.  Wrote a sample F# app that seems to work just fine:

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>

  <ItemGroup>
    <Compile Include="Program.fs" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="GraphQL.AspNetCore3" Version="2.1.0" />
    <PackageReference Include="GraphQL.Server.Ui.Playground" Version="5.2.1" />
    <PackageReference Include="GraphQL.SystemTextJson" Version="5.1.1" />
  </ItemGroup>

</Project>
```

With

```fsharp
open System
open Microsoft.AspNetCore.Builder
open Microsoft.Extensions.Hosting
open GraphQL
open GraphQL.AspNetCore3
open GraphQL.MicrosoftDI
open GraphQL.SystemTextJson
open GraphQL.Types
open System.Collections.Generic
open Microsoft.Extensions.DependencyInjection

type TransportationTypeEnumGraphType () as x =
    inherit EnumerationGraphType ()

    do
        x.Name <- "TransportationTypeEnum"
        x.Description <- "Transportation type"

        x.Add (name = "PASSENGER_TRANSFER", value = "PASSENGER_TRANSFER", description = "Passenger transfer")
        x.Add (name = "CREW_SAFEGUARDING", value = "CREW_SAFEGUARDING", description = "Crew safeguarding")

type DataClass () =
    member val Name : string = null with get, set
    [<InputType(typeof<ListGraphType<TransportationTypeEnumGraphType>>)>] member val Types : List<string> = null with get, set

type MyQuery () =
    static member Hello (stuff : DataClass) : string =
        String.Join(",", stuff.Types)


[<EntryPoint>]
let main args =
    let builder = WebApplication.CreateBuilder(args)
    builder.Services.AddGraphQL(fun b ->
        b.AddAutoSchema<MyQuery>() |> ignore
        b.AddGraphTypes() |> ignore
        b.AddSystemTextJson() |> ignore) |> ignore
    let app = builder.Build()

    //verify it initializes correctly
    app.Services.GetRequiredService<ISchema>().Initialize()

    app.UseGraphQL() |> ignore
    app.UseGraphQLPlayground("/") |> ignore

    app.Run()

    0 // Exit code

```

Query

```gql
query ($arg:[TransportationTypeEnum])
{
  hello(stuff:{
    name:"abc"
    types:$arg
  })
}
```

Variables:

```json
{
  "arg":["PASSENGER_TRANSFER"]
}
```

**Comment by Shane32** [comment]
FYI, per the GraphQL specs, a list of one item can be represented by the item itself, so that's why this works also:

```json
{
  "arg":"PASSENGER_TRANSFER"
}
```

**Comment by draganjovanovic1** [comment]
Ok. Will try to break it. Yes, you are right for non array representation, it worked fine previously as well.

**Comment by sungam3r** [comment]
GraphQL.NET package has embedded pdbs and supports SourceLink. You may enable SourceLink from your IDE and drill into the internals.

**Comment by draganjovanovic1** [comment]
@Shane32 missed you question on `typeName` definition. I use naming convention to differentiate GQL mapped classes from others by adding `GraphType` suffix. `TransportationTypeEnum` is `TransportationTypeEnumGraphType` so `typeName` just trims `GraphType` to have nicer name in GQL schema definition. 

Here it is:
```fsharp
    let typeName (x: 'a :> GraphType) = x.GetType().Name.Replace("GraphType", "")
```

I haven't tried your example yet but first thing I notice is that it uses `System.Text.Json` and I use `Newtonsoft.Json`. Do you run tests against both JSON libraries?

**Comment by Shane32** [comment]
Yes we run comprehensive tests against both libraries

**Comment by Shane32** [comment]
You can change the sample to use Newtonsoft by changing one line of code and the dependency just to be sure.

**Comment by draganjovanovic1** [comment]
Sure, thanks. Will investigate and post my findings here.

**Comment by draganjovanovic1** [comment]
@Shane32 it was a problem with deserialization of GQL request. I host graphql endpoint in Giraffe web api and it has it's own serialization settings so `InputsJsonConverter` wasn't actually used. Variables enum collection was actually parsed as a collection of `JToken`s instead of collection of `string`s. Totally ignored the removal of `ToInputs` extension method which I used in v4. 🤦‍♂️

![image](https://user-images.githubusercontent.com/4232435/164461428-dd0fec5b-5d18-452b-9343-17d18385b4b8.png)

```csharp
        public override object? ParseValue(object? value) => value switch
        {
            string s => Values.FindByName(s)?.Value ?? ThrowValueConversionError(value),
            null => null,
            _ => ThrowValueConversionError(value) // here it goes as it wasn't string
        };
```

I fixed it by using `InputsJsonConverter`.
